app/src/main/java/com/ubergeek42/WeechatAndroid/fragments/BufferFragment.java

```diff
@@ -351,14 +360,124 @@ public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
+        if (RelayConnection.weechatVersion >= 0x2090000 && !tcInProgress) {
+            tabCompleteRelayStrategy();
+        } else {
+            tabCompleteLocalStrategy();
+        }
+    }
+
+    private String tcViaRelayBaseString; // the entire original input
+    private int tcViaRelayIndex; // index used when looping over tcViaRelayCompletionResults
+    private String[] tcViaRelayCompletionResults;
+    private boolean tcViaRelayAddSpace; // to add a space after completion
+    private int tcViaRelayPosStart; // beginning of section to be replaced by one of the suggestions
+    private int tcViaRelayPosEnd; // end of said section in tcViaRelayBaseString
+    // tcViaRelayBaseString.subString(tcViaRelayPosStart, tcViaRelayPosEnd + 1)
+    private String tcViaRelayBaseWord;
+    private int tcViaRelaySelectionIndex; // where the cursor was when tab was pressed
+
+    @MainThread private void tabCompleteRelayStrategy() {
+        // disable the uiInput, re-enable it after the future has been completed
+        uiInput.setFocusable(false);
+
+        if (tcViaRelayInProgress) {
+            tcViaRelayIndex = (tcViaRelayIndex + 1) % tcViaRelayCompletionResults.length;
+            StringBuilder st = new StringBuilder(tcViaRelayBaseString);
+            st.replace(tcViaRelayPosStart, tcViaRelayPosEnd + 1,
+                    tcViaRelayCompletionResults[tcViaRelayIndex] +
+                            (tcViaRelayAddSpace ? " " : ""));
+            uiInput.setText(st.toString());
+            uiInput.setSelection(tcViaRelaySelectionIndex +
+                    tcViaRelayCompletionResults[tcViaRelayIndex].length()
+                    - tcViaRelayBaseWord.length() + (tcViaRelayAddSpace ? 1 : 0));
+            uiInput.setFocusable(true);
+            uiInput.setFocusableInTouchMode(true);
+            uiInput.setPressed(true);
+            tcViaRelayInProgress = true;
+            return;
+        }
+
+        // in the case of a selection, autocomplete from the left border
+        tcViaRelaySelectionIndex = uiInput.getSelectionStart();
+        tcViaRelayBaseString = uiInput.getText().toString();
+
+        CompletableFuture<RelayObject[]> completionPromise =
+                SendMessageWithPromiseEvent.fire("completion",
+                        "completion 0x%x %d %s", buffer.pointer,
+                        tcViaRelaySelectionIndex, tcViaRelayBaseString);
+
+        // https://weechat.org/files/doc/devel/weechat_relay_protocol.en.html#command_completion
+        new Thread(() -> {
+            try {
+                RelayObject[] objects = completionPromise.get(10, SECONDS);
+                // the documention mentions that "the input string; if not given, completion is
+                // performed on an empty string"
+                // empirically, the Hdata is empty if completion is performed on an empty string
+                if (((Hdata) objects[0]).getCount() > 0) {
```

Comment:
it would be nice not to have the “low-level” code related to `RelayObject`s here maybe, although it would probably complicate things if it's put elsewhere

Reply:
You know what, now that I'm adhering to the existing `RelayMessageHandler` pattern in `BufferList.java` it isn't so complicated anymore; I should go about doing this (i.e. calling `tcViaRelayCompletionResults = item.getItem("list").asArray().asStringArray()` and resolving the promise with a POJO called `TabCompletionResonse` or something